### PR TITLE
Update HTTPClient_AddHeader() to not accept invalid characters in the field and value

### PR DESCRIPTION
**Pull request by sarenameas**
HTTPCLient_AddHeader() is updated to check for carriage return, linefeed, and colons in the input.
CBMC proofs are updated to support this.

### CODE REVIEW

File: source/core_http_client.c [hunk 1]
```diff
@@ -1186,6 +1254,8 @@ static HTTPStatus_t addHeader( HTTPRequestHeaders_t * pRequestHeaders,
     char * pBufferCur = NULL;
     size_t toAddLen = 0u;
     size_t backtrackHeaderLen = pRequestHeaders->headersLen;
+    const uint8_t isField = 1;
+    const uint8_t isValue = 0;
```
Comment (yourslab) [inline source/core_http_client.c:1258]: This private function is called throughout many public-facing functions, so I'm wondering if we could avoid the extra stack usage by having the above as an enum or macro.

Comment (sarenameas) [reply source/core_http_client.c:1258]: sure

File: test/cbmc/proofs/HTTPClient_InitializeRequestHeaders/Makefile [hunk 2]
```diff
@@ -38,13 +38,22 @@ REMOVE_FUNCTION_BODY += __CPROVER_file_local_core_http_client_c_httpParserOnHead
 REMOVE_FUNCTION_BODY += __CPROVER_file_local_core_http_client_c_httpParserOnMessageBeginCallback
 REMOVE_FUNCTION_BODY += __CPROVER_file_local_core_http_client_c_httpParserOnMessageCompleteCallback
 
+# These functions are removed and replace with stubs that check that the
+# destination parameter is writeable and that the source parameter is readable.
+# This increases the run-time of the proof. This is safe to do for this proof
```
Comment (sarenameas) [inline test/cbmc/proofs/HTTPClient_InitializeRequestHeaders/Makefile:43]:
```suggestion
# This decreases the run-time of the proof. This is safe to do for this proof
```

File: test/cbmc/stubs/httpHeaderStrncpy.c [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ * HTTPClient_InitializeRequestHeaders run much faster. This stub checks if, for
+ * the input copy length, the destination and source are valid accessible
+ * memory.
+ */
+
+#include <string.h>
+#include <stdint.h>
+
+void * httpHeaderStrncpy( void * pDest,
+                          const void * pSrc,
```
Comment (yourslab) [inline test/cbmc/stubs/httpHeaderStrncpy.c:36]: Why are we using `void *` instead?

Comment (sarenameas) [reply test/cbmc/stubs/httpHeaderStrncpy.c:36]: i copied and pasted and forgot to change the type thanks

**Review by yourslab — APPROVED** [review]

**Review by n9wxu — APPROVED** [review]

**Pull request merged**